File: inJava/app/src/main/java/com/barrygithub/rickandmortyapp/di/AppModule.java
package com.barrygithub.rickandmortyapp.di;

import android.app.Application;

import androidx.room.Room;

import com.barrygithub.rickandmortyapp.common.Constants;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.MyDAO;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.MyDatabase;
import com.barrygithub.rickandmortyapp.data.RemoteDatasource.ApiClient;
import com.barrygithub.rickandmortyapp.data.Repository;
import com.barrygithub.rickandmortyapp.data.RepositoryImpl;
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractor;
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractorImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 3/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 **/
@Module
@InstallIn(SingletonComponent.class)
public class AppModule {

    @Singleton
    @Provides
    public MyDAO databaseBuilder(Application app) {
        return Room.databaseBuilder(app.getApplicationContext(),
                        MyDatabase.class,
                        Constants.DATABASE_NAME
                ).fallbackToDestructiveMigration()
                 .build().myDAO()
                ;
    }

    @Singleton
    @Provides
    public Repository repositoryProvide(ApiClient apiClient, MyDAO db){return new RepositoryImpl(apiClient, db);}

    @Singleton
    @Provides
    public UseCaseInteractor interactorProvide(Repository repository) {return new UseCaseInteractorImpl(repository);}

}

File: inKotlin/app/src/main/java/com/barrygithub/rickandmortyapp/di/AppModule.kt
package com.barrygithub.rickandmortyapp.di

import android.app.Application
import androidx.room.Room
import com.barrygithub.rickandmortyapp.common.Constants
import com.barrygithub.rickandmortyapp.data.localDatasource.MyDAO
import com.barrygithub.rickandmortyapp.data.localDatasource.MyDatabase
import com.barrygithub.rickandmortyapp.data.remoteDatasource.ApiClient
import com.barrygithub.rickandmortyapp.data.remoteDatasource.Repository
import com.barrygithub.rickandmortyapp.data.remoteDatasource.RepositoryImpl
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractor
import com.barrygithub.rickandmortyapp.domain.UseCaseInteractorImpl
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 14/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 *
 **/
@Module
@InstallIn(SingletonComponent::class)
class AppModule {

    @Singleton
    @Provides
    fun databaseBuilder(app:Application):MyDAO{
        return Room.databaseBuilder(
            app.applicationContext,
            MyDatabase::class.java,
            Constants.DATABASE_NAME
        ).fallbackToDestructiveMigration()
            .build()
            .myDAO()
    }

    @Singleton
    @Provides
    fun repositoryProvides(apiClient:ApiClient, db:MyDAO):Repository = RepositoryImpl(apiClient,db)

    @Provides
    fun useCaseProvides(repository: Repository):UseCaseInteractor = UseCaseInteractorImpl(repository)


}